src/expressions/OrderByExpression.ts

```diff
@@ -0,0 +1,343 @@
+						dynamicContexts.push(iteratorResult.value);
+						iteratorResult = dynamicContextIterator.next(IterationHint.NONE);
+					}
+
+					// Evaluate order specs. Limited to only one order spec for now.
+					const evaluatedOrderSpecs: ISequence[] = dynamicContexts.map(
+						(dynamicContextForEvaluation) =>
+							orderSpec.expression.evaluate(
+								dynamicContextForEvaluation,
+								executionParameters
+							)
+					);
+
+					// Atomize
+					// Atomization is applied to the result of the expression in each orderspec. If the result of atomization is neither a single atomic value nor an empty sequence, a type error is raised [err:XPTY0004].
+					const atomizedSequences: ISequence[] = evaluatedOrderSpecs.map(
+						(evaluatedOrderSpec) => atomize(evaluatedOrderSpec, executionParameters)
+					);
+					if (atomizedSequences.find((val) => !val.isEmpty() && !val.isSingleton())) {
+						throw errXPTY0004('Order by only accepts empty or singleton sequences');
+					}
+
+					// Switch to values instead of sequences as we now know we're dealing with singletons only.
+					values = atomizedSequences.map((seq) => seq.first());
+
+					// Casting values
+					// If the value of an orderspec has the dynamic type xs:untypedAtomic, it is cast to the type xs:string.
+					values = values.map((value) => {
+						if (value === null) {
+							return value;
+						}
+
+						if (isSubtypeOf('xs:untypedAtomic', value.type)) {
+							return castToType(value, 'xs:string');
+						}
+
+						return value;
+					});
+
+					// If the resulting sequence contains values that are instances of more than one primitive type, then:
+					// 1. If each value is an instance of one of the types xs:string or xs:anyURI, then all values are cast to type xs:string.
+					// 2. If each value is an instance of one of the types xs:decimal or xs:float, then all values are cast to type xs:float.
+					// 3. If each value is an instance of one of the types xs:decimal, xs:float, or xs:double, then all values are cast to type xs:double.
+					// 4. Otherwise, a type error is raised [err:XPTY0004].
+					const firstPrimitiveType: string | null = getFirstPrimitiveType(values);
+
+					if (firstPrimitiveType) {
+						const areOfSamePrimitiveType: boolean = values.every((value) =>
+							value ? getPrimitiveTypeName(value.type) === firstPrimitiveType : true
+						);
+
+						if (!areOfSamePrimitiveType) {
+							values = tryCastToSingleType(values);
+						}
+					}
+
+					// Sort values
+					const numberOfValues: number = values.length;
+					indices = values.map((_value, index) => index);
+					for (let startIndex: number = 0; startIndex < numberOfValues; startIndex++) {
+						if (startIndex + 1 === numberOfValues) {
+							continue;
+						}
+
+						for (let i: number = startIndex; i >= 0; i--) {
+							const firstItemIndex = i;
+							const secondItemIndex = i + 1;
+
+							if (secondItemIndex === numberOfValues) {
+								continue;
```

Comment:
When does this happen? It is uncovered in coveralls: https://coveralls.io/builds/34525433/source?filename=src%2Fexpressions%2FOrderByExpression.ts#L40